DigitalLearningSolutions.Data.Tests/Services/StoreAspServiceTests.cs

```diff
@@ -0,0 +1,703 @@
+                .And(p => p.Completed = null)
+                .And(p => p.CustomisationId = DefaultCustomisationId)
+                .And(p => p.IsProgressLocked = false)
+                .TheFirst(1).With(p => p.RemovedDate = DateTime.UtcNow)
+                .TheNext(1).With(p => p.Completed = DateTime.UtcNow)
+                .TheNext(1).With(p => p.CustomisationId = DefaultCustomisationId + 100)
+                .TheNext(1).With(p => p.IsProgressLocked = true)
+                .Build();
+            A.CallTo(() => courseDataService.GetDelegateCoursesInfo(A<int>._)).Returns(progressRecords);
+
+            // When
+            var result = storeAspService.GetProgressAndValidateInputsForStoreAspAssess(
+                1,
+                2,
+                DefaultDelegateId,
+                DefaultCustomisationId
+            );
+
+            // Then
+            using (new AssertionScope())
+            {
+                result.validationResponse.Should().Be(TrackerEndpointResponse.StoreAspAssessException);
+                result.progress.Should().BeNull();
+                A.CallTo(() => courseDataService.GetDelegateCoursesInfo(DefaultDelegateId))
+                    .MustHaveHappenedOnceExactly();
+            }
+        }
+
+        [Test]
+        public void
+            GetProgressAndValidateInputsForStoreAspAssess_returns_StoreAspAssessException_if_multiple_valid_progress_records_obtained()
+        {
+            // Given
+            var progressRecords = Builder<DelegateCourseInfo>.CreateListOfSize(5).All()
+                .With(p => p.RemovedDate = null)
+                .And(p => p.Completed = null)
+                .And(p => p.CustomisationId = DefaultCustomisationId)
+                .And(p => p.IsProgressLocked = false)
+                .TheFirst(1).With(p => p.RemovedDate = DateTime.UtcNow)
+                .TheNext(1).With(p => p.Completed = DateTime.UtcNow)
+                .TheNext(1).With(p => p.CustomisationId = DefaultCustomisationId + 100)
+                .Build();
+            A.CallTo(() => courseDataService.GetDelegateCoursesInfo(A<int>._)).Returns(progressRecords);
+
+            // When
+            var result = storeAspService.GetProgressAndValidateInputsForStoreAspAssess(
+                1,
+                2,
+                DefaultDelegateId,
+                DefaultCustomisationId
+            );
+
+            // Then
+            using (new AssertionScope())
+            {
+                result.validationResponse.Should().Be(TrackerEndpointResponse.StoreAspAssessException);
+                result.progress.Should().BeNull();
+                A.CallTo(() => courseDataService.GetDelegateCoursesInfo(DefaultDelegateId))
+                    .MustHaveHappenedOnceExactly();
+            }
+        }
+
+        [Test]
+        public void
+            GetProgressAndValidateCommonInputsForStoreAspProgressEndpoints_returns_progress_record_and_null_exception_when_all_is_valid()

```

Comment:
Instead of `null_exception`, I'd prefer either `no_exception` or `null_validationResponse`, since I immediately thought: "Is a null exception like Java's null pointer exception?"

Note: there are numerous other tests with this naming convention both in this file and in TrackerActionServiceTests